hedera-node/src/main/java/com/hedera/services/contracts/execution/HederaTracer.java

```diff
@@ -37,25 +37,158 @@
-    }
-
-    @Override
-    public void traceAccountCreationResult(
-            final MessageFrame frame, final Optional<ExceptionalHaltReason> haltReason) {
-        frame.setExceptionalHaltReason(haltReason);
-    }
+@Singleton
+public class HederaTracer implements HederaOperationTracer {
+
+	private final List<SolidityAction> allActions;
+	private final Deque<SolidityAction> currentActionsStack;
+	private boolean areActionSidecarsEnabled;
+
+	@Inject
+	public HederaTracer() {
+		this.currentActionsStack = new ArrayDeque<>();
+		this.allActions = new ArrayList<>();
+		this.areActionSidecarsEnabled = false;
+	}
+
+	@Override
+	public void reset(final boolean areActionSidecarsEnabled) {
+		this.currentActionsStack.clear();
+		this.allActions.clear();
+		this.areActionSidecarsEnabled = areActionSidecarsEnabled;
+	}
+
+	@Override
+	public void traceExecution(MessageFrame frame, ExecuteOperation executeOperation) {
+		if (areActionSidecarsEnabled && currentActionsStack.isEmpty()) {
+			trackActionFor(frame, 0);
+		}
+
+		executeOperation.execute();
+
+		if (areActionSidecarsEnabled) {
+			final var frameState = frame.getState();
+			if (frameState != State.CODE_EXECUTING) {
+				if (frameState == State.CODE_SUSPENDED) {
+					final var nextFrame = frame.getMessageFrameStack().peek();
+					trackActionFor(nextFrame, nextFrame.getMessageFrameStack().size() - 1);
+				} else {
+					finalizeActionFor(currentActionsStack.pop(), frame, frameState);
+				}
+			}
+		}
+	}
+
+	@Override
+	public void tracePrecompileResult(final MessageFrame frame, final ContractActionType type) {
+		if (areActionSidecarsEnabled) {
+			final var lastAction = currentActionsStack.pop();
+			// specialize the call type - precompile or system (Hedera precompile contracts)
+			lastAction.setCallType(type);
+			// we have to null out recipient account and set recipient contract
+			lastAction.setRecipientAccount(null);
+			lastAction.setRecipientContract(EntityId.fromAddress(frame.getContractAddress()));
+
+			finalizeActionFor(lastAction, frame, frame.getState());
+		}
+	}
+
+	private void trackActionFor(final MessageFrame frame, final int callDepth) {
+		// code can be empty when calling precompiles too, but we handle
+		// that in tracePrecompileCall, after precompile execution is completed
+		final var isCallToAccount = Code.EMPTY.equals(frame.getCode());
+		final var isTopLevelEVMTransaction = callDepth == 0;
+		final var action = new SolidityAction(
+				toContractActionType(frame.getType()),
+				isTopLevelEVMTransaction ? EntityId.fromAddress(frame.getOriginatorAddress()) : null,
+				!isTopLevelEVMTransaction ? EntityId.fromAddress(frame.getSenderAddress()) : null,
+				frame.getRemainingGas(),
+				frame.getInputData().toArray(),
+				isCallToAccount ? EntityId.fromAddress(frame.getContractAddress()) : null,
+				!isCallToAccount ? EntityId.fromAddress(frame.getContractAddress()) : null,
+				frame.getValue().toLong(),
+				callDepth);
```

Comment:
Are we OK with having the fields `input`, `gas`, `amount`, `recipient` for initial frames both in the transaction body and in the first contract action or should we avoid the duplication?